### DISCUSSION TITLE
Enter (Return Key) as Tab to navigation

### DISCUSSION THREAD

**Discussion by marcosgerene** [AvaloniaUI/Avalonia#12414] [Q&A] [answered]
Hi everyone,

I am working on a Desktop application and I want to do something like this (it works on WPF for me):

http://madprops.org/blog/enter-to-tab-as-an-attached-property/

How can I do this on Avalonia (desktop)?

---

Edit:

I have made this and kind of worked:

```

public static void FocusNext(this Window window)
{
	var current = window.FocusManager.GetFocusedElement();

	if (current != null)
		MoveNext(current);
}

public static void FocusNext(this InputElement element)
{
	var focusManager = TopLevel.GetTopLevel(element).FocusManager;

	var current = focusManager.GetFocusedElement();

	if (current != null)
		MoveNext(current);
}

private static void MoveNext(IInputElement current)
{
	var next = KeyboardNavigationHandler.GetNext(current, NavigationDirection.Next);

	while (true)
	{
		if (next == null)
			break;

		if (next is ComboBox || next is TextBox || next is CheckBox || next is NumericUpDown)
			break;

		next = KeyboardNavigationHandler.GetNext(next, NavigationDirection.Next);
	}

	next?.Focus(NavigationMethod.Directional, KeyModifiers.None);

	if (next != null && next is not InputElement)
		MoveNext(next);
}

public static void MoveNext_KeyDown(object? sender, KeyEventArgs e)
{
	if (e.Key != Key.Return)
		return;

	if (sender == null)
		return;

	e.Handled = true;

	if (sender is Window window)
		window.FocusNext();

	else if (sender is InputElement element)
		element.FocusNext();  
}
```

But I am facing some problems:

1. Once the focus is on a CheckBox or ComboBox I can't make this fire, it kind of does not allow me to rewrite the KeyDown return, as it is handled =true.
2. Is there any parameter as "IsTabStop", cause I want to ignore some fields

**Comment by IanRawley** [comment, marked as answer]
1.  You have three options here:  In code behind use the AddHandler() method to add a handler that catches tunneled keydown events (the equivalent of a KeyDownPreview), in XAML use a behaviour (see https://github.com/AvaloniaUI/Avalonia.Xaml.Behaviors) to catch the event on tunneling.  Alternatively subclass Check/ComboBox and override OnKeyDown.
2. It's an attached property, KeyboardNavigation.IsTabStop

**Comment by marcosgerene** [comment]
@IanRawley 

>You have three options here: In code behind use the AddHandler() method to add a handler that catches tunneled keydown events (the equivalent of a KeyDownPreview), in XAML use a behaviour (see https://github.com/AvaloniaUI/Avalonia.Xaml.Behaviors) to catch the event on tunneling. Alternatively subclass Check/ComboBox and override OnKeyDown.

I am pretty new on Avalonia, so I did not get how to do that. 

I don't want to be abusive, but can you gimme an example in code?

> It's an attached property, KeyboardNavigation.IsTabStop

Ok, it worked on Checkbox, ComboBox and TextBox. How to do it in the button (up and down) of a NumericUpDown?

**Reply by IanRawley** [reply]
That's fine, I'm pretty new myself.
For AddHandler it looks like this:
```
public PropertyView()
    {
        InitializeComponent();
        
        PropertyList.AddHandler<KeyEventArgs>(KeyDownEvent, PropertyListKeyDown, RoutingStrategies.Bubble, true);
    }
```

Where PropertyList is ListView with x:Name=PropertyList in the XAML.

Behaviors is messier, and I couldn't figure out how to get the pre-designed stuff to catch keydown on tunneled and call a code-behind method so built my own.  Someone else can probably do that better.

For the NumericUpDown, this is where your particular theme will come in.  I think by default Avalonia uses the Fluent theme, but I found Simple quicker for my use cases.  In the Simple theme a NumericUpDown is actually a TextBox wrapped in a ButtonSpinner, so you'll need to use a style to change IsTabStop on the individual parts.  I've done it app wide for ButtonSpinners due to how I'm using NumericUpDowns with the following in App.axaml, but what works for you will be more specific to your use case.

```
<Style Selector="ButtonSpinner">
	<Setter Property="IsTabStop" Value="False"/>
</Style>
```

**Comment by marcosgerene** [comment]
The AddHandler thing kind of worked. I will have to do some ajusts on my code, but it worked pretty well, thanks!

>For the NumericUpDown, this is where your particular theme will come in. I think by default Avalonia uses the Fluent theme, but I found Simple quicker for my use cases. In the Simple theme a NumericUpDown is actually a TextBox wrapped in a ButtonSpinner, so you'll need to use a style to change IsTabStop on the individual parts. I've done it app wide for ButtonSpinners due to how I'm using NumericUpDowns with the following in App.axaml, but what works for you will be more specific to your use case.
```
<Style Selector="ButtonSpinner">
	<Setter Property="IsTabStop" Value="False"/>
</Style>
```

I am using the default theme (FluentTheme). It is part of a series of studying projects. The style you sent worked for the NumericUpDown itself, but not for the buttons.

**Reply by IanRawley** [reply]
I forgot I've also got NumericUpDowns styled to not show the spinner, so that's probably why that's enough for me.  You'd probably want to add that style to your NumericUpDown and/or add a style targeting RepeatButton (as that's what a Fluent ButtonSpinner is made from) with the same setter.

**Reply by marcosgerene** [reply]
It could work in my case. How to make make then hidden/colapsed?

**Reply by IanRawley** [reply]
NumericUpDown has a ShowButtonSpinner property you can set to false.
https://docs.avaloniaui.net/docs/next/reference/controls/numericupdown

**Reply by marcosgerene** [reply]
Thank you very much!
